lib/src/switch.dart

```diff
@@ -62,8 +62,8 @@ extension SwitchLatest<T> on Stream<Stream<T>> {
       var outerStreamDone = false;
 
       final outerSubscription = listen(
-          (innerStream) {
-            innerSubscription?.cancel();
+          (innerStream) async {
+            await innerSubscription?.cancel();
```

Comment:
Not sure it's worth waiting for the `cancel` to complete if you don't handle an error anyway.

If `cancel` returns an error future, then that error becomes an uncaught error. The `listen` callbacks are not supposed to throw, and if they do, the error just becomes uncaught.
The `await` means that the following code doesn't run in that case, which it did before.

If the `cancel` completes normally, you wait for it, and that *might* be reasonable, but also might just be a delay.

Since you don't pause the outer stream while waiting, this code may be reentered before it's done, if another stream arrives while waiting for the `cancel`. It'll then try to cancel the same `innerSubscription` twice, then eventually listen on two (or more) streams, and not be able to cancel all of them.

Maybe something like:
```dart
try {
  outerSubscription.pause();
  await innerSubscription?.cancel();
} finally {
  the rest
  outerSubscription.resume();
}
```
(or use `.whenComplete` directly instead of `async` + `try`/`finally`).

Or, alternatively, if you don't want to pause the outer stream, you need some state management. If a new stream arrives while waiting for an old stream to cancel, you just override the "pending stream" with the new stream, and never listen to to the stream that got superseeded before it was started (unless you actually need/want to wait `listen`+`cancel` the stream to not leave the creator dangling - shouldn't be necessary, but sometimes is).

That's a larger rewrite (but not unexpected for something involving multiple streams or other async operations).

Reply:
Also, if the returned stream is paused during an inner-stream cancel, you still listen to the next inner-stream afterwards, and it isn't paused.